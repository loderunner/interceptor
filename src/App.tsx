import { Box, Stack } from '@mui/material'
import * as React from 'react'
import { useEffect, useReducer } from 'react'

import type Protocol from 'devtools-protocol'

type TargetInfo = chrome.debugger.TargetInfo
type Debuggee = chrome.debugger.Debuggee
type RequestPausedEvent = Protocol.Fetch.RequestPausedEvent

const App = () => {
  const [requests, pushRequest] = useReducer(
    (requests: string[], req: string) => [...requests, req],
    []
  )

  const onEvent = (target: Debuggee, method: string, params?: unknown) => {
    console.log('debugger event', method, params)
    if (method === 'Fetch.requestPaused') {
      const event: RequestPausedEvent = params as RequestPausedEvent
      const d = new Date()
      pushRequest(
        `${d.toISOString()} - ${event.request.method} ${event.request.url}`
      )
      chrome.debugger.sendCommand(target, 'Fetch.continueRequest', {
        requestId: event.requestId,
      })
    }
  }

  const getTargets = async () => {
    return new Promise<TargetInfo[]>((resolve) =>
      chrome.debugger.getTargets((targets) => resolve(targets))
    )
  }

  const initDebugger = async () => {
    const targets = await getTargets()
    for (const t of targets) {
      if (t.tabId === chrome.devtools.inspectedWindow.tabId) {
        const debuggee: chrome.debugger.Debuggee = { targetId: t.id }
        await chrome.debugger.attach(debuggee, '1.3')
        chrome.debugger.onEvent.addListener(onEvent)
        await chrome.debugger.sendCommand(debuggee, 'Fetch.enable', {})
      }
    }
  }

  useEffect(() => {
    initDebugger()
  }, [])

  return (
    <>
      <Box sx={{ typography: 'h1' }}>Hello World!</Box>
      <Stack>
        {requests.map((req, i) => (
          <div key={i}>{req}</div>
        ))}
      </Stack>
    </>
  )
}

export default App
