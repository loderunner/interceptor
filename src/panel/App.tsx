import { Box, Stack } from '@mui/material'
import * as React from 'react'
import { useEffect, useReducer } from 'react'

import * as Requests from './requests'

const App = () => {
  const [requests, pushRequest] = useReducer(
    (requests: Requests.Request[], req: Requests.Request) => [...requests, req],
    []
  )

  useEffect(() => {
    Requests.listen()
    const unsubscribe = Requests.subscribe((req) => {
      pushRequest(req)
    })

    return () => {
      unsubscribe()
      Requests.unlisten()
    }
  }, [])

  return (
    <>
      <Box sx={{ typography: 'h1' }}>Hello World!</Box>
      <Stack>
        {requests.map((request, i) => (
          <div key={i}>
            {request.method} {request.url}
          </div>
        ))}
      </Stack>
    </>
  )
}

export default App
