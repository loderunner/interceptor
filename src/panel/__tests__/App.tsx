import { shallow } from 'enzyme'
import React from 'react'

import App from '../App'

describe('[App]', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot()
  })
})
