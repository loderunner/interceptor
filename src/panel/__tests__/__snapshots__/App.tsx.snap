// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[App] matches snapshot 1`] = `
<Fragment>
  <ForwardRef(Box)
    sx={
      Object {
        "typography": "h1",
      }
    }
  >
    Hello World!
  </ForwardRef(Box)>
  <ForwardRef(Stack) />
</Fragment>
`;
